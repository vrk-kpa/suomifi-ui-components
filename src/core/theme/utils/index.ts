export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
