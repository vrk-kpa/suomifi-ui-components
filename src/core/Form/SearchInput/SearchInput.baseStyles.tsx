import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import {
  input,
  containerIEFocus,
  font,
  fixInternalMargins,
} from '../../theme/reset';
import { MarginProps, buildSpacingCSS } from '../../theme/utils/spacing';

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${font(theme)('bodyText')}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
  ${fixInternalMargins()}
  width: 290px;

  & .fi-search-input {
    &_wrapper {
      width: 100%;
      min-width: 105px;
      display: inline-block;

      & .fi-search-input_label--visible {
        margin-bottom: ${theme.spacing.xs};
      }

      & .fi-search-input_statusText--has-content {
        margin-top: ${theme.spacing.xxs};
      }
    }

    &_functionality-container {
      position: relative;
    }

    &_input-element-container {
      background-color: ${theme.colors.whiteBase};
      color: ${theme.colors.blackBase};
      ${containerIEFocus(theme)}

      &:focus-within {
        position: relative;
        box-shadow: ${theme.shadows.actionElementBoxShadow};
        ${theme.focuses.highContrastFocus}
        &:after {
          ${theme.focuses.absoluteFocus}
          top: -3px;
          bottom: -3px;
          right: -3px;
          left: -3px;
        }
        > input:focus {
          outline: none;
        }
      }
      width: 100%;
      height: 40px;
      box-sizing: border-box;
      border: 1px solid ${theme.colors.depthDark3};
      border-radius: ${theme.radiuses.basic};
    }

    &_input {
      ${input(theme)}
      padding-top: ${theme.spacing.insetS};
      padding-bottom: ${theme.spacing.insetS};
      width: calc(100% - 24px);
      min-width: 65px;
      border: 0;
      min-height: 36px;
      margin-top: 1px;
      margin-bottom: 1px;
      &::placeholder {
        font-style: italic;
        color: ${theme.colors.depthDark2};
        opacity: 1;
      }
      &::-ms-clear {
        display: none;
        width: 0;
        height: 0;
      }
      &::-ms-reveal {
        display: none;
        width: 0;
        height: 0;
      }
      &::-webkit-search-decoration,
      &::-webkit-search-cancel-button,
      &::-webkit-search-results-button,
      &::-webkit-search-results-decoration {
        display: none;
      }
      appearance: none;
    }

    &_button {
      position: absolute;
      top: 0;
      display: flex;
      justify-content: center;
      align-items: center;
      box-sizing: border-box;
      outline: none;
      box-shadow: none;
      cursor: default;
      pointer-events: none;
      &-clear {
        right: 40px;
        clip: rect(0 0 0 0);
        height: 1px;
        width: 1px;
        margin: -1px;
        padding: 0;
        border: 0;
        overflow: hidden;

        &:focus {
          ${theme.focuses.highContrastFocus}
        }
        &-icon {
          width: 12px;
          height: 12px;
          & .fi-icon-base-fill {
            fill: ${theme.colors.highlightDark1};
          }
        }
      }

      &-search {
        right: 0px;
        height: 40px;
        width: 40px;
        border-radius: 0 ${theme.radiuses.basic} ${theme.radiuses.basic} 0;
        border: 0;
        &-icon {
          width: 18px;
          height: 18px;
          & .fi-icon-base-fill {
            fill: ${theme.colors.depthDark1};
          }
        }
      }
    }
  }

  &.fi-search-input--full-width {
    width: 100%;
  }

  &.fi-search-input--error {
    & .fi-search-input_input-element-container {
      border: 1px solid ${theme.colors.alertBase};
      border-right: 0;
    }
    & .fi-search-input_button-search {
      border: 1px solid ${theme.colors.alertBase};
      border-left: 0;
    }
  }

  &.fi-search-input--not-empty {
    & .fi-search-input_input-element-container {
      width: calc(100% - 40px);
      border-radius: ${theme.radiuses.basic} 0 0 ${theme.radiuses.basic};
      border-right: 0;
    }

    & .fi-search-input_button {
      cursor: pointer;
      pointer-events: all;
    }

    & .fi-search-input_button-clear {
      clip: auto;
      overflow: visible;
      height: 20px;
      width: 20px;
      margin: 10px;

      & .fi-icon-base-fill {
        @media (forced-colors: active) {
          fill: ButtonText;
        }
      }
    }

    & .fi-search-input_button-search {
      background: ${theme.gradients.highlightBaseToHighlightDark1};

      /* Support for high contrast mode */
      @media (forced-colors: active) {
        border: solid 1px ButtonBorder;
      }

      &:focus {
        &:after {
          ${theme.focuses.absoluteFocus}
        }

        ${theme.focuses.highContrastFocus}/* For high contrast mode */
      }
      &:hover {
        background: ${theme.gradients.highlightLight1ToHighlightBase};
      }
      &:active {
        background-color: ${theme.colors.highlightDark1};
      }
      & .fi-search-input_button-search-icon .fi-icon-base-fill {
        fill: ${theme.colors.whiteBase};

        @media (forced-colors: active) {
          fill: ButtonText;
        }
      }
    }
  }
`;
