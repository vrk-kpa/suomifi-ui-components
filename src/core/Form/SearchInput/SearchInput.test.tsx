import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { axeTest } from '../../../utils/test';

import { SearchInput, SearchInputProps } from './SearchInput';

const TestSearchInput = (props: Partial<SearchInputProps> = {}) => {
  const { labelText, clearButtonLabel, searchButtonLabel, ...passProps } =
    props;
  return (
    <SearchInput
      data-testid="searchinput"
      labelText={labelText || 'Test search input'}
      clearButtonLabel={clearButtonLabel || 'Clear'}
      searchButtonLabel={searchButtonLabel || 'Search'}
      {...passProps}
    />
  );
};

describe('snapshot', () => {
  it('should have matching default structure', () => {
    const searchInputRendered = render(TestSearchInput());
    const { container } = searchInputRendered;
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('props', () => {
  describe('default structure, with default props', () => {
    it('should have baseClassName', () => {
      const { container } = render(TestSearchInput());
      expect(container.firstChild).toHaveClass('fi-search-input');
    });

    it('should have <input type="search"> HTML-element with correct class ', () => {
      const { getByRole } = render(TestSearchInput());
      const inputElement = getByRole('searchbox');
      expect(inputElement).toHaveClass('fi-search-input_input');
      expect(inputElement).toHaveAttribute('type', 'search');
    });

    it('should have label text with correct class and for id', () => {
      const { getByText } = render(TestSearchInput({ id: 'test-id' }));
      const label = getByText('Test search input').closest('label');
      expect(label).toHaveClass('fi-label_label-span');
      expect(label).toHaveAttribute('for', 'test-id');
    });

    it(
      'should not have basic accessibility issues',
      axeTest(TestSearchInput()),
    );
  });

  describe('className', () => {
    it('should have the given custom classname ', () => {
      const { container } = render(
        TestSearchInput({ className: 'custom-classname' }),
      );
      expect(container.firstChild).toHaveClass('custom-classname');
    });
  });

  describe('margin', () => {
    it('should have margin style from margin prop', () => {
      const { container } = render(
        <SearchInput
          labelText=""
          clearButtonLabel=""
          searchButtonLabel=""
          margin="xs"
        />,
      );
      expect(container.firstChild).toHaveStyle('margin: 10px');
    });

    it('should have margin prop overwritten by style prop', () => {
      const { container } = render(
        <SearchInput
          labelText=""
          clearButtonLabel=""
          searchButtonLabel=""
          margin="xs"
          style={{ margin: 2 }}
        />,
      );
      expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
    });
  });

  describe('onBlur', () => {
    test('should notice when leaving area', () => {
      const mockOnBlur = jest.fn();
      const { getByRole } = render(TestSearchInput({ onBlur: mockOnBlur }));
      const inputElement = getByRole('searchbox');
      fireEvent.blur(inputElement);
      expect(mockOnBlur).toBeCalledTimes(1);
    });
  });

  describe('onChange', () => {
    it('should notice change and have the given text', () => {
      const mockOnChange = jest.fn();
      const { getByRole } = render(TestSearchInput({ onChange: mockOnChange }));
      const inputElement = getByRole('searchbox') as HTMLTextAreaElement;
      fireEvent.change(inputElement, { target: { value: 'abc' } });
      expect(mockOnChange).toBeCalledTimes(1);
      expect(inputElement.value).toBe('abc');
    });
  });

  describe('onSearch', () => {
    it('should trigger onSearch callback', () => {
      const mockOnSearch = jest.fn();
      const { getAllByRole } = render(
        TestSearchInput({
          onSearch: mockOnSearch,
          defaultValue: 'Some test value',
        }),
      );
      const searchButton = getAllByRole('button')[1];
      fireEvent.click(searchButton);
      expect(mockOnSearch).toBeCalledTimes(1);
    });
  });

  describe('onClear', () => {
    it('should trigger onChange and clear input value', () => {
      const mockOnChange = jest.fn();
      const { getAllByRole } = render(
        TestSearchInput({
          onChange: mockOnChange,
          defaultValue: 'Some test value',
        }),
      );
      const clearButton = getAllByRole('button')[0];
      fireEvent.click(clearButton);
      expect(mockOnChange).toBeCalledTimes(1);
      const inputElement = getAllByRole('searchbox')[0] as HTMLInputElement;
      expect(inputElement.value).toBe('');
    });
  });

  describe('labelMode', () => {
    it('hidden: should hide visually-hidden classname   ', () => {
      const { getByText } = render(
        TestSearchInput({
          labelMode: 'hidden',
          labelText: 'To be hidden',
        }),
      );
      const label = getByText('To be hidden');
      expect(label).toHaveClass('fi-visually-hidden');
    });
  });

  describe('visualPlaceholder', () => {
    it('should have the given text as attribute', () => {
      const { getByRole } = render(
        TestSearchInput({
          visualPlaceholder: 'Search...',
        }),
      );
      const inputElement = getByRole('searchbox') as HTMLInputElement;
      expect(inputElement).toHaveAttribute('placeholder', 'Search...');
    });
  });

  describe('statusText', () => {
    it('should have element and correct classname for it', () => {
      const { getByText } = render(
        TestSearchInput({ statusText: 'EROR EROR' }),
      );
      const statusText = getByText('EROR EROR');
      expect(statusText).toHaveClass('fi-status-text');
    });
  });

  describe('status', () => {
    it('should have error classname', () => {
      const { container } = render(
        TestSearchInput({ status: 'error', statusText: 'EROR EROR' }),
      );
      expect(container.firstChild).toHaveClass('fi-search-input--error');
    });
  });

  describe('name', () => {
    it('has the given name attribute', () => {
      const { getByRole } = render(TestSearchInput({ name: 'test-name' }));
      const inputElement = getByRole('searchbox') as HTMLInputElement;
      expect(inputElement.name).toBe('test-name');
    });
  });

  describe('aria', () => {
    it('input should have provided aria-describedby and attribute referencing status text', () => {
      const { getByRole } = render(
        TestSearchInput({
          id: 'test-id-aria',
          'aria-describedby': 'describedby-id',
          statusText: 'Test status text',
        }),
      );
      const inputElement = getByRole('searchbox') as HTMLInputElement;
      expect(inputElement).toHaveAttribute(
        'aria-describedby',
        'test-id-aria-statusText describedby-id',
      );
    });

    it('icons should have aria-hidden attribute set to true', () => {
      const { container } = render(TestSearchInput());
      const svgList = container.getElementsByTagName('svg');
      for (let i = 0; i < svgList.length; i += 1) {
        expect(svgList[i]).toHaveAttribute('aria-hidden', 'true');
      }
    });
  });
  describe('debounce', () => {
    it('delays the running of onChange by the given time', () => {
      jest.useFakeTimers();
      const mockOnChange = jest.fn();
      const searchInput = (
        <SearchInput
          labelText="Debounced search"
          clearButtonLabel="clear"
          searchButtonLabel="search"
          debounce={1000}
          onChange={mockOnChange}
        />
      );
      const { getByRole } = render(searchInput);

      const inputElement = getByRole('searchbox') as HTMLInputElement;
      fireEvent.change(inputElement, { target: { value: 'new value' } });
      expect(mockOnChange).not.toBeCalled();
      jest.advanceTimersByTime(1000);
      expect(mockOnChange).toBeCalledTimes(1);
      expect(inputElement.value).toBe('new value');
    });
  });

  describe('ref', () => {
    it('ref is forwarded to input', () => {
      const ref = React.createRef<HTMLInputElement>();

      render(
        <SearchInput
          labelText="Debounced search"
          clearButtonLabel="clear"
          searchButtonLabel="search"
          ref={ref}
        />,
      );

      expect(ref.current?.tagName).toBe('INPUT');
    });
  });
});

describe('states', () => {
  describe('empty input', () => {
    it('should not have accessible buttons', () => {
      const { queryAllByRole } = render(TestSearchInput({ defaultValue: '' }));
      const buttons = queryAllByRole('button');
      expect(buttons).toHaveLength(0);
    });

    it('should not react to enter before adding text to input', () => {
      const mockOnSearch = jest.fn();
      const { getByRole, getAllByRole } = render(
        TestSearchInput({
          onSearch: mockOnSearch,
          defaultValue: '',
        }),
      );
      const inputElement = getByRole('searchbox') as HTMLTextAreaElement;
      fireEvent.keyPress(inputElement, {
        key: 'Enter',
        code: 13,
        charCode: 13,
      });
      expect(mockOnSearch).toBeCalledTimes(0);
      fireEvent.change(inputElement, { target: { value: 'abc' } });
      fireEvent.keyPress(inputElement, {
        key: 'Enter',
        code: 13,
        charCode: 13,
      });
      expect(mockOnSearch).toBeCalledTimes(1);
      const searchButton = getAllByRole('button')[1];
      fireEvent.click(searchButton);
      expect(mockOnSearch).toBeCalledTimes(2);
    });
  });
});
