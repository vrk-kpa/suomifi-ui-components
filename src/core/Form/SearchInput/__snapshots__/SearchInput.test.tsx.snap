// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should have matching default structure 1`] = `
.c6 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c6:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c6::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c6::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c6::-webkit-inner-spin-button {
  height: auto;
}

.c6::-webkit-outer-spin-button {
  height: auto;
}

.c6::before,
.c6::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c3::before,
.c3::after {
  box-sizing: border-box;
}

.c5 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
}

.c5::-webkit-input-placeholder {
  color: inherit;
  opacity: 0.54;
}

.c5::before,
.c5::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c8 {
  position: absolute;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
}

.c9 {
  vertical-align: baseline;
}

.c9.fi-icon {
  display: inline-block;
}

.c9 .fi-icon-base-fill {
  fill: currentColor;
}

.c9 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c9.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c9.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c10 {
  vertical-align: baseline;
}

.c10.fi-icon {
  display: inline-block;
}

.c10 .fi-icon-base-fill {
  fill: currentColor;
}

.c10 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c10.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c10.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c11 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(0,0%,13%);
  font-size: 14px;
  line-height: 20px;
}

.c11.fi-status-text {
  display: block;
}

.c11.fi-status-text.fi-status-text--error {
  color: hsl(3,59%,43%);
}

.c11.fi-status-text .fi-icon {
  vertical-align: middle;
  -webkit-transform: translateY(-0.1em);
  -ms-transform: translateY(-0.1em);
  transform: translateY(-0.1em);
  margin-right: 5px;
}

.c4.fi-label .fi-label_label-span {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  display: inline-block;
  vertical-align: middle;
  color: hsl(0,0%,13%);
}

.c4.fi-label .fi-label_label-span .fi-label_optional-text {
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c4.fi-label .fi-label_label-span .fi-tooltip {
  margin-left: 6px;
}

.c7 {
  position: relative;
  cursor: pointer;
  pointer-events: all;
  height: 20px;
  width: 20px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7:focus {
  outline: 3px solid transparent;
}

.c7:focus:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c7.fi-input-clear-button {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c7.fi-input-clear-button .fi-input-clear-button_icon {
  width: 16px;
  height: 16px;
}

.c7.fi-input-clear-button .fi-input-clear-button_icon .fi-icon-base-fill {
  fill: hsl(212,63%,37%);
}

.c7.fi-input-clear-button[disabled],
.c7.fi-input-clear-button:disabled {
  cursor: not-allowed;
}

.c7.fi-input-clear-button[disabled] .fi-input-clear-button_icon .fi-icon-base-fill,
.c7.fi-input-clear-button:disabled .fi-input-clear-button_icon .fi-icon-base-fill {
  fill: hsl(202,7%,67%);
}

.c1 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  width: 290px;
}

.c1 .fi-label,
.c1 .fi-hint-text,
.c1 .fi-status-text {
  margin: 0;
}

.c1 .fi-search-input_wrapper {
  width: 100%;
  min-width: 105px;
  display: inline-block;
}

.c1 .fi-search-input_wrapper .fi-search-input_label--visible {
  margin-bottom: 10px;
}

.c1 .fi-search-input_wrapper .fi-search-input_statusText--has-content {
  margin-top: 5px;
}

.c1 .fi-search-input_functionality-container {
  position: relative;
}

.c1 .fi-search-input_input-element-container {
  background-color: hsl(0,0%,100%);
  color: hsl(0,0%,13%);
  width: 100%;
  height: 40px;
  box-sizing: border-box;
  border: 1px solid hsl(201,7%,58%);
  border-radius: 2px;
}

.c1 .fi-search-input_input-element-container > input:focus {
  outline-color: hsl(196,77%,44%);
  outline-width: 2px;
  outline-offset: 2px;
  outline-style: solid;
}

.c1 .fi-search-input_input-element-container:focus-within > input:focus {
  outline: none;
}

.c1 .fi-search-input_input-element-container:focus-within {
  position: relative;
  box-shadow: 0 1px 2px 0 hsla(214,100%,24%,0.1) inset;
  outline: 3px solid transparent;
}

.c1 .fi-search-input_input-element-container:focus-within:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
  top: -3px;
  bottom: -3px;
  right: -3px;
  left: -3px;
}

.c1 .fi-search-input_input-element-container:focus-within > input:focus {
  outline: none;
}

.c1 .fi-search-input_input {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  min-width: 245px;
  max-width: 100%;
  padding: 8px 16px;
  border: 1px solid hsl(202,7%,80%);
  border-radius: 2px;
  line-height: 1;
  padding-top: 6px;
  padding-bottom: 6px;
  width: calc(100% - 24px);
  min-width: 65px;
  border: 0;
  min-height: 36px;
  margin-top: 1px;
  margin-bottom: 1px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c1 .fi-search-input_input::-webkit-input-placeholder {
  font-style: italic;
  color: hsl(201,7%,46%);
  opacity: 1;
}

.c1 .fi-search-input_input::-moz-placeholder {
  font-style: italic;
  color: hsl(201,7%,46%);
  opacity: 1;
}

.c1 .fi-search-input_input:-ms-input-placeholder {
  font-style: italic;
  color: hsl(201,7%,46%);
  opacity: 1;
}

.c1 .fi-search-input_input::placeholder {
  font-style: italic;
  color: hsl(201,7%,46%);
  opacity: 1;
}

.c1 .fi-search-input_input::-ms-clear {
  display: none;
  width: 0;
  height: 0;
}

.c1 .fi-search-input_input::-ms-reveal {
  display: none;
  width: 0;
  height: 0;
}

.c1 .fi-search-input_input::-webkit-search-decoration,
.c1 .fi-search-input_input::-webkit-search-cancel-button,
.c1 .fi-search-input_input::-webkit-search-results-button,
.c1 .fi-search-input_input::-webkit-search-results-decoration {
  display: none;
}

.c1 .fi-search-input_button {
  position: absolute;
  top: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  outline: none;
  box-shadow: none;
  cursor: default;
  pointer-events: none;
}

.c1 .fi-search-input_button-clear {
  right: 40px;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
}

.c1 .fi-search-input_button-clear:focus {
  outline: 3px solid transparent;
}

.c1 .fi-search-input_button-clear-icon {
  width: 12px;
  height: 12px;
}

.c1 .fi-search-input_button-clear-icon .fi-icon-base-fill {
  fill: hsl(212,63%,37%);
}

.c1 .fi-search-input_button-search {
  right: 0px;
  height: 40px;
  width: 40px;
  border-radius: 0 2px 2px 0;
  border: 0;
}

.c1 .fi-search-input_button-search-icon {
  width: 18px;
  height: 18px;
}

.c1 .fi-search-input_button-search-icon .fi-icon-base-fill {
  fill: hsl(202,7%,40%);
}

.c1.fi-search-input--full-width {
  width: 100%;
}

.c1.fi-search-input--error .fi-search-input_input-element-container {
  border: 1px solid hsl(3,59%,43%);
  border-right: 0;
}

.c1.fi-search-input--error .fi-search-input_button-search {
  border: 1px solid hsl(3,59%,43%);
  border-left: 0;
}

.c1.fi-search-input--not-empty .fi-search-input_input-element-container {
  width: calc(100% - 40px);
  border-radius: 2px 0 0 2px;
  border-right: 0;
}

.c1.fi-search-input--not-empty .fi-search-input_button {
  cursor: pointer;
  pointer-events: all;
}

.c1.fi-search-input--not-empty .fi-search-input_button-clear {
  -webkit-clip: auto;
  clip: auto;
  overflow: visible;
  height: 20px;
  width: 20px;
  margin: 10px;
}

.c1.fi-search-input--not-empty .fi-search-input_button-search {
  background: linear-gradient(0deg,hsl(212,63%,37%) 0%,hsl(212,63%,45%) 100%);
}

.c1.fi-search-input--not-empty .fi-search-input_button-search:focus {
  outline: 3px solid transparent;
}

.c1.fi-search-input--not-empty .fi-search-input_button-search:focus:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c1.fi-search-input--not-empty .fi-search-input_button-search:hover {
  background: linear-gradient(0deg,hsl(212,63%,45%) 0%,hsl(212,63%,49%) 100%);
}

.c1.fi-search-input--not-empty .fi-search-input_button-search:active {
  background-color: hsl(212,63%,37%);
}

.c1.fi-search-input--not-empty .fi-search-input_button-search .fi-search-input_button-search-icon .fi-icon-base-fill {
  fill: hsl(0,0%,100%);
}

@media (forced-colors:active) {
  .c1.fi-search-input--not-empty .fi-search-input_button-clear .fi-icon-base-fill {
    fill: ButtonText;
  }
}

@media (forced-colors:active) {
  .c1.fi-search-input--not-empty .fi-search-input_button-search {
    border: solid 1px ButtonBorder;
  }
}

@media (forced-colors:active) {
  .c1.fi-search-input--not-empty .fi-search-input_button-search .fi-search-input_button-search-icon .fi-icon-base-fill {
    fill: ButtonText;
  }
}

<div
  class="c0 c1 fi-search-input"
>
  <span
    class="c2 fi-search-input_wrapper"
  >
    <div
      class="c3 c4 fi-search-input_label--visible fi-label"
    >
      <label
        class="c2 fi-label_label-span"
        for="1"
      >
        Test search input
      </label>
    </div>
    <div
      class="c0 fi-search-input_functionality-container"
    >
      <div
        class="c0 fi-search-input_input-element-container"
      >
        <input
          aria-invalid="false"
          class="c5 fi-search-input_input"
          data-testid="searchinput"
          id="1"
          role="searchbox"
          type="search"
          value=""
        />
      </div>
      <button
        class="c6 fi-input-clear-button c7 fi-search-input_button fi-search-input_button-clear"
        hidden=""
        tabindex="-1"
        type="button"
      >
        <span
          class="c2 c8 fi-visually-hidden"
        >
          Clear
        </span>
        <svg
          aria-hidden="true"
          class="fi-icon c9 fi-input-clear-button_icon"
          focusable="false"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class="fi-icon-base-fill"
            d="M4.75 3.3 12 10.55l7.25-7.25a1.026 1.026 0 0 1 1.45 1.45L13.45 12l7.25 7.25a1.026 1.026 0 0 1-1.45 1.45L12 13.45 4.75 20.7a1.026 1.026 0 0 1-1.45-1.45L10.55 12 3.3 4.75A1.026 1.026 0 0 1 4.75 3.3Z"
            fill="#222"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <button
        class="c6 fi-search-input_button fi-search-input_button-search"
        hidden=""
        tabindex="-1"
        type="button"
      >
        <span
          class="c2 c8 fi-visually-hidden"
        >
          Search
        </span>
        <svg
          aria-hidden="true"
          class="fi-icon c10 fi-search-input_button-search-icon"
          focusable="false"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class="fi-icon-base-fill"
            d="M9 0c4.962 0 9 4.038 9 9.001a8.955 8.955 0 0 1-1.971 5.614l7.678 7.678a.999.999 0 1 1-1.414 1.414l-7.678-7.678A8.957 8.957 0 0 1 9 18c-4.962 0-9-4.037-9-8.999C0 4.038 4.038 0 9 0Zm0 2C5.14 2 2 5.141 2 9.001 2 12.86 5.14 16 9 16s7-3.14 7-6.999C16 5.141 12.86 2 9 2Z"
            fill="#222"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
    <span
      aria-atomic="true"
      aria-live="assertive"
      class="c2 c11 fi-status-text"
      id="1-statusText"
    />
  </span>
</div>
`;
